Splitting `_invariant` Lemmas

cwaldm opened the pull request:
Fixing https://github.com/REPROSEC/dolev-yao-star-extrinsic/issues/80.

I also added "get_state => state_was_set" Lemmas (core, tagged, typed), which are often useful.
(For this, I needed the `last_entry_exists` Lemma in `Trace.Base`)

TWal commented on src/core/DY.Core.Trace.Base.fst:
```diff
@@ -254,6 +254,20 @@ val event_at_grows:
+       let Snoc _ last = tr in
+       event_exists tr last
+    ))
+    [SMTPat (Snoc? tr)]
```
I am genuinely surprised about this SMT pattern, it works without having to write `assert(Snoc? tr)`?!

cwaldm replied:
It works well in `Snoc` cases of `match`es.
Do you suggest a different pattern?

TWal replied:
No, if it works well that is fine! I never wrote a pattern like that so I didn't know how it would behave.

TWal commented on src/core/DY.Core.Trace.Manipulation.fst:
```diff
@@ -486,6 +534,56 @@ let rec get_state_aux_state_invariant #invs prin sess_id tr =
+  (ensures (
+    let (opt_content, tr_out) = get_state prin sess_id tr in
+    tr == tr_out
+    ))
```
Nit (indentation)
```suggestion
  ))
```

TWal commented on src/core/DY.Core.Trace.Manipulation.fst:
```diff
@@ -486,6 +534,56 @@ let rec get_state_aux_state_invariant #invs prin sess_id tr =
+    match get_state_aux prin sess_id tr with
+    | None -> True
+    | Some content -> 
+          state_was_set tr prin sess_id content
```
Nit (indentation)
```suggestion
      state_was_set tr prin sess_id content
```

TWal replied:
Why not, but it is not [the style advocated by CONTRIBUTING.md](https://github.com/REPROSEC/dolev-yao-star-extrinsic/blob/main/CONTRIBUTING.md#inside-a-match), and it is not consistent with the rest of the codebase (which consistently uses indentations of two spaces without consideration for alignment). The rationale being that we want to avoid cases such as
```fstar
match x with
| SomeReallyLongAndExplicitNameThatOccupiesALotOfSpace foo ->
                                                       this will easily go too far on the right
| Bar bar ->
      this is not indented as the code two lines above
```

For this reason I will always prefer the two-space indentation style, but it's not worthy of having a debate on that, so do as you prefer. But if you want to keep the indent-to-align style, please choose between
```fstar
match x with
| Some y ->
      foo
```
and
```fstar
match x with
| Some y ->
       foo
```
as the two styles exist in this PR.

TWal commented on src/core/DY.Core.Trace.Manipulation.fst:
```diff
@@ -486,6 +534,56 @@ let rec get_state_aux_state_invariant #invs prin sess_id tr =
+      match opt_content with
+      | None -> True
+      | Some content -> 
+             state_was_set tr prin sess_id content
```
Nit (indentation)
```suggestion
        state_was_set tr prin sess_id content
```

TWal commented on src/lib/state/DY.Lib.State.Typed.fst:
```diff
@@ -186,17 +200,50 @@ val set_state_invariant:
+  (ensures (
+    let (opt_content, tr_out) = get_state #a prin sess_id tr in
+    tr == tr_out
+    ))
```
nit (indentation)
```suggestion
  ))
```

TWal commented on src/lib/state/DY.Lib.State.Tagged.fst:
```diff
@@ -276,20 +289,56 @@ val set_tagged_state_invariant:
+         tagged_state_was_set tr tag prin sess_id content
+      )
+    )
+  )
+  )
```
nit
```suggestion
  ))
```

TWal commented on src/lib/state/DY.Lib.State.Tagged.fst:
```diff
@@ -276,20 +289,56 @@ val set_tagged_state_invariant:
+  (ensures (
+    let (opt_content, tr_out) = get_tagged_state tag prin sess_id tr in
+    tr == tr_out 
+    ))
```
nit (indentation)
```suggestion
  ))
```

TWal commented on src/lib/state/DY.Lib.State.PrivateKeys.fst:
```diff
@@ -183,19 +195,30 @@ val get_private_key_invariant:
+  (ensures (
+    let (opt_private_key, tr_out) = compute_public_key prin sess_id pk_type tr in
+    tr_out == tr
+    ))
```
nit (indentation)
```suggestion
  ))
```

TWal commented on src/lib/state/DY.Lib.State.PKI.fst:
```diff
@@ -117,6 +117,18 @@ val install_public_key_invariant:
+  (ensures (
+    let (opt_public_key, tr_out) = get_public_key prin sess_id pk_type who tr in
+    tr_out == tr
+    ))
```
nit (indentation)
```suggestion
  ))
```

TWal commented on src/lib/state/DY.Lib.State.Map.fst:
```diff
@@ -237,6 +237,21 @@ let add_key_value_invariant #invs #key_t #value_t #mt mpred prin sess_id key val
+  (ensures (
+    let (opt_value, tr_out) = find_value #_ #value_t prin sess_id key tr in
+    tr_out == tr
+    ))
```
nit (indentation)
```suggestion
  ))
```

TWal commented on src/core/DY.Core.Trace.Manipulation.fst:
```diff
@@ -186,6 +205,17 @@ let recv_msg i =
+  Lemma
+  (ensures (
+    let (opt_msg, tr_out) = recv_msg i tr in
+    tr_out == tr))
```
nit
```suggestion
    tr_out == tr
  ))
```

TWal commented on src/core/DY.Core.Trace.Manipulation.fst:
```diff
@@ -115,6 +115,15 @@ let add_event e tr =
   norm_spec [zeta; delta_only [`%prefix]] (prefix #label);
   ((), Snoc tr e)
 
+val add_event_event_exists:
+  e:trace_event -> tr:trace ->
+  Lemma
+  (ensures (
+    let ((), tr_out) = add_event e tr in
+    event_exists tr_out e
+  ))
+let add_event_event_exists e tr = ()
```
It seems this is not used anywhere, and it is not a lemma we expect the user to use, so maybe we don't need it and can remove it?

cwaldm replied:
On the other hand, it doesn't hurt to have it.
Not for the user, but for us at some later point maybe. When we add other trace entries.
But I don't really mind. So either way is fine for me.

TWal replied:
Ok, why not keeping it then (let's say it's a useful sanity check anyway)

TWal reviewed (commented):
This PR looks good, I posted many comments but most of them are formatting nit-picking on the lemma statements.

Indeed there was missing lemmas that `get_state` implies `state_was_set`. I notice we also have lemmas that `get_state` implies `state_pred`, maybe instead of the latter we could use lemmas (with SMT patterns) that `state_was_set` implies `state_pred`?
It means that going from `get_state` to `state_pred` requires two SMT triggers instead of one, but I guess it is okay, and because `set_state` implies `state_was_set` this would also trigger the new lemma but I guess it is also okay.

What do you think about it?

cwaldm commented on src/lib/state/DY.Lib.State.Tagged.fst:
```diff
@@ -276,20 +289,56 @@ val set_tagged_state_invariant:
+  Lemma
+  (ensures (
+    let (opt_content, tr_out) = get_tagged_state tag prin sess_id tr in
+    tr == tr_out /\ (
```
Correct!
I overlooked that. Will remove it from here.

cwaldm commented:
We already have all "state_was_set => state_pred" lemmas.
Together with the new "get_state => state_was_set" lemmas, we can indeed just remove the "get_state => state_pred" lemmas, if you want.

TWal commented on src/lib/state/DY.Lib.State.Typed.fst:
```diff
@@ -186,42 +200,49 @@ val set_state_invariant:
       match opt_content with
       | None -> True
-      | Some content -> (
-        spred.pred tr prin sess_id content
-      )
-    )
+      | Some content -> 
+             state_was_set tr prin sess_id content
   ))
```
GitHub will not let me propose a change, but I think the indentation should be
```fstar
   let (opt_content, tr_out) = get_state #a #ls prin sess_id tr in
    match opt_content with
    | None -> True
    | Some content -> 
```
i.e. the match shouldn't be indented, this made sense when it was enclosed in `tr == tr_out /\ ( ... )` but it is not the case anymore

TWal commented on src/lib/state/DY.Lib.State.Typed.fst:
```diff
@@ -186,17 +200,50 @@ val set_state_invariant:
+      match opt_content with
+      | None -> True
+      | Some content -> 
+             state_was_set tr prin sess_id content
```
superseded by another comment on a more recent commit

cwaldm commented:
I hope, I addressed all layout issues. (Choosing 2 spaces indent for the `Some` cases.)
(I also removed a now unnecessary `invariant` lemma for `get_state_aux`.)
